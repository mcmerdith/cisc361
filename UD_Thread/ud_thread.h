/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);
