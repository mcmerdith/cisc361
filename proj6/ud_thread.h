/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);

/**
 * semaphore
 */

// Create a new semaphore pointed to by sp with a count value of sem_count.
int sem_init(sem_t **sp, int sem_count);

// Current thread does a wait (P) on the specified semaphore.
void sem_wait(sem_t *sp);

// Current thread does a signal (V) on the specified semaphore.
// Follow the Mesa semantics (p. 9 of OSTEP Chapter 30 Condition Variables) where the thread that signals continues,
// and the first waiting (blocked) thread (if there is any) becomes ready.
void sem_signal(sem_t *sp);

// Free any memory related to the specified semaphore.
// If there are queued TCBs, they are moved to the ready queue, and this may imply that the semantics of the application might be incorrect.
void sem_destroy(sem_t **sp);

typedef void sem_t;