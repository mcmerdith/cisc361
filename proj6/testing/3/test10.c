/*
 * Test Program #10 - Critical Sections with Yields
 */

#include <stdio.h>
#include <stdlib.h>
#include "ud_thread.h"

sem_t *mutex;
int resource = 0;

void read_function(int val)
{
    printf("I am thread %d...\n", val);

    sem_wait(mutex);
    printf("thread %d entering 1st CS...\n", val);
    resource = 11;
    printf("in thread: resource is %d\n", resource);
    t_yield();
    resource = 22;
    printf("in thread: resource is %d\n", resource);
    printf("thread %d leaving 1st CS...\n", val);
    sem_signal(mutex);

    sem_wait(mutex);
    printf("thread %d entering 2nd CS...\n", val);
    resource = 55;
    printf("in thread: resource is %d\n", resource);
    t_yield();
    resource = 66;
    printf("in thread: resource is %d\n", resource);
    printf("thread %d leaving 2nd CS...\n", val);
    sem_signal(mutex);

    printf("thread %d: t_terminate()\n", val);
    t_terminate();
}

int main(void)
{
    t_init();
    sem_init(&mutex, 1);

    t_create(read_function, 123, 1);

    t_yield();

    sem_wait(mutex);
    printf("main thread entering CS...\n");
    resource = 33;
    printf("in main: resource is %d\n", resource);
    t_yield();
    resource = 77;
    printf("in main: resource is %d\n", resource);
    printf("main thread leaving CS...\n");
    sem_signal(mutex);

    printf("main thread gives up CPU...\n");
    t_yield();
    t_yield();

    printf("in main: sem_destroy() and t_shutdown()\n");
    sem_destroy(&mutex);
    t_shutdown();

    return 0;
}