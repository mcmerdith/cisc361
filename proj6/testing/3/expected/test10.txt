I am thread 123...
thread 123 entering 1st CS...
in thread: resource is 11
in thread: resource is 22
thread 123 leaving 1st CS...
main thread entering CS...
in main: resource is 33
in main: resource is 77
main thread leaving CS...
main thread gives up CPU...
thread 123 entering 2nd CS...
in thread: resource is 55
in thread: resource is 66
thread 123 leaving 2nd CS...
thread 123: t_terminate()
in main: sem_destroy() and t_shutdown()